feat(avm)!: WIP ALU FDIV

MirandaWood opened the pull request:
## a / b = c, v.2: electric boogaloo

Adds field division opcode `FDIV` (somewhat following [vm1](https://github.com/AztecProtocol/aztec-packages/blame/bf7a99d8d8e3faa00c9395a17acf7f92b171a7a1/barretenberg/cpp/pil/avm/main.pil#L328)). Given fields `a` and `b`, shows that `a / b = c` by constraining `a = c * b`.

I re-use the existing `DIV` relation by forcing the remainder to be zero, and add the case of a non-FF input as a tag error. Also uses the existing divide by zero error plus tests!

MirandaWood commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -311,6 +286,47 @@ sel_is_u128 * sel_op_div * (1 - sel_err) * a_hi * b_hi = 0;
+
+sel_op_fdiv * (1 - sel_op_fdiv) = 0;
+
+// For FDIV, we reuse the same main relation for non-u128 DIV (ALU_FDIV_DIV_NON_U128) but constrain that there is no remainder, stored in helper1.
+// => The relation shows that b * c = a for FF
+
+// DIV & FDIV - shared ops
+// #[OP_ID_CHECK] ensures selectors are mutually exclusive:
+pol DIV_OPS = sel_op_div + sel_op_fdiv;
+
+pol commit sel_div_0_err;
+sel_div_0_err * (1 - sel_div_0_err) = 0;
+
+// We prove that sel_div_0_err == 1 <==> DIV_OPS == 1 && b == 0:
+//      sel_div_0_err == 1 <==> b == 0:         ib * (sel_div_0_err * (1 - b_inv) + b_inv) + sel_div_0_err - 1 = 0
+//      sel_div_0_err == 1 => DIV_OPS == 1:  sel_div_0_err * ( sel_div_0_err - DIV_OPS ) = 0
+
+// Proof:
+// ==>:
+// Assume sel_div_0_err == 1. Then #[DIV_0_ERR] becomes: DIV_OPS * ib + 1 - DIV_OPS = 0
+// We note that DIV_OPS cannot be zero, otherwise we get 1 == 0.
+// Since DIV_OPS is boolean, DIV_OPS == 1 and ib == 0
+// <== :
+// Assume DIV_OPS == 1 && ib == 0, we get for #[DIV_0_ERR]: sel_div_0_err - 1 + sel_div_0_err * (sel_div_0_err - 1) = 0
+// sel_div_0_err being boolean the second additive term is zero and therefore sel_div_0_err - 1 == 0.
+pol commit b_inv;
+#[DIV_0_ERR]
+DIV_OPS * (ib * (sel_div_0_err * (1 - b_inv) + b_inv) + sel_div_0_err - 1) + sel_div_0_err * (sel_div_0_err - DIV_OPS) = 0;
+
+// DIV & FDIV - non u128
+
+// FDIV note: I'm using (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff) to avoid having to set and check sel_is_u128 for FDIV, since we know it's FF => non u128
+pol DIV_OPS_NON_U128 = DIV_OPS * (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff);
+
+// Show a - remainder = b * c
+// Note: Since a, b, c, and remainder are under 64 bits (or remainder == 0 for FF), we are protected against a - remainder underflowing,
+// i.e. b * c (< 128bits) cannot equal underflowed a - remainder (> 253 bits)
+#[ALU_FDIV_DIV_NON_U128]
+DIV_OPS_NON_U128 * (1 - sel_err) * (ib * ic - (ia - sel_op_div * helper1)) = 0;
+
```
Most of this is not new - just moved the sections of code for non-u128 `DIV` down and added cases for `FDIV`

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/common/instruction_spec.cpp:
```diff
@@ -452,6 +452,13 @@ const std::unordered_map<ExecutionOpCode, ExecInstructionSpec> EXEC_INSTRUCTION_
+      { .num_addresses = 3,
+        .gas_cost = { .opcode_gas = AVM_FDIV_BASE_L2_GAS, .base_da = 0, .dyn_l2 = 0, .dyn_da = 0 },
+        .register_info = RegisterInfo()
+                             .add_inputs({ /*a*/ RegisterInfo::ANY_TAG,
```
Hmm in the FDIV case we could just require both tags to be FF here?

Then whether you actually do FF tag checking in the ALU or not for FDIV is a consideration of that component. The nicest, of course, is to do it. But if it makes things simpler you could consider
(1) asserting the tag: that is, things would be unprovable if not FF, but in practice you should never get there from execution
(2) not checking at all: not the best from a design PoV, for a conceptually sensical ALU but worth considering if massively simplifying things

Lets see what @IlyasRidhuan & @jeanmon think

MirandaWood replied:
Oh true! I was kind of on autopilot here. In case it's not super clear in the code, this PR checks the tag `is_ff` in `alu.pil` and turns on `sel_tag_err` if it's not. I'm happy to incorporate any of the options :)

IlyasRidhuan replied:
IIUC I dont think you can do (1) - not because it's not a good idea, but because at the execution trace you don't know this is an FDIV (only the ALU knows that) so it can't be asserted in registers.

IlyasRidhuan replied:
i don't mind (2) actually. I assume noir/brillig only generates this for field types, but i dont see why it couldnt just disregard the tag (not sure the total savings IMO)

fcarreiro replied:
To clarify, I meant doing FF in the register info (then execution can check that) but then option 1 and 2 are for the ALU itself, inside the ALU.

jeanmon replied:
My take is that #[EXEC_SPEC_READ] is a capability which already implements this restriction, i.e., mandating a tag for a register. I would be really a pity not to use it. We "unfortunately cannot" use it for other alu operations where "only a given tag value is disallowed."

I am suggesting to enforce the "ff tag" for FDIV in execution and do 2) but we comment in alu.pil that this is a precondition enforced in execution.pil.

@IlyasRidhuan This mechanism here from #[EXEC_SPEC_READ] lookup is based on the execution opcode and does not require a binary opcode selector (which you are right is not present in execution.pil)

jeanmon commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -124,10 +126,10 @@ pol commit sel_tag_err;
 // 2. Mismatched tags for inputs a and b for all opcodes apart from TRUNC
 // 1 is handled by checking FF_TAG_ERR in TAG_ERR_CHECK and 2 is handled in AB_TAGS_CHECK.
-pol FF_TAG_ERR = (sel_op_div + sel_op_not) * sel_is_ff;
+pol FF_TAG_ERR = (sel_op_div + sel_op_not) * sel_is_ff + sel_op_fdiv * IS_NOT_FF;
```
I can live with the current plan but I find it more readable (similar performance) to add a single relation:
`sel_of_fdiv * (ic_tag - FF_TAG_COINSTANT) = 0;`
and not change anything for: CHECK_TAG_FF, EXPECTED_C_TAG and FF_TAG_ERR.

MirandaWood replied:
I think I'm missing something here - is this suggestion to add a new relation which checks the `c_tag` is `FF`?
(The highlighted bit of code above only related to input tags with recoverable tag errors using `sel_tag_err`).
I agree it could be more readable, but I also don't want to clutter the file with lots of new lines! Maybe I could change `EXPECTED_C_TAG` to be `FF` for `fdiv` rather than `ia_tag`?

jeanmon replied:
What I was originally suggesting is to remove the changes pertaining to: CHECK_TAG_FF, EXPECTED_C_TAG and FF_TAG_ERR.
and adding a dedicated relation enforcing `c_tag == FF`.

Now, I agree that your suggestion to change `EXPECTED_C_TAG` to be `FF` for `fdiv` is a better idea.
If we do that and assuming that "register mechanism" enforces that `ia_tag` and `ib_tag` are `FF`, I think you can revert the changes made in `CHECK_TAG_FF` and `FF_TAG_ERR`

MirandaWood replied:
Ah ok now I understand! Sorry, I thought you meant keep _my_ changes as is for the three vars. I will change `EXPECTED_C_TAG` to be `FF`.
As to the others, I thought we wanted to keep checks both in execution (i.e. register mechanism) and in `alu.pil` (same as other comment re facilitating unit tests), but happy to change these!

jeanmon replied:
The whole consensus was to take the path which leads to the easiest/efficient constraints. From what I see, removing the changes for `CHECK_TAG_FF` and `FF_TAG_ERR` lead to simpler constraints. 
Maybe we can have a little comment mentioning that execution constrains ia_tag == ib_tag == FF.

fcarreiro replied:
I thought the consensus we reached was that keeping the ALU self-consistent was more important than optimization _unless_ (of course) the gains were "big enough". Of course "big enough" is subjective, but my take is that if mostly existing "infra" then then gain would have to be > 4 columns and 3 or 4 subrelations.

jeanmon replied:
Yeah self-consistency was mentioned (You might be right on the consensus @fcarreiro). However, we have a much more important non self-consistency issue which is that we rely on the range check in write memory operation for ic in different places (ALU and possibly other gadgets.) To be coherent we should address this one as well but might be a bit more annoying to solve if we have to add a range-check lookup and adding it in simulation.

jeanmon replied:
@MirandaWood Sorry for this. Let us follow self-consistency as @fcarreiro mentioned. Regarding the extra range-check for write, I will open a slack thread.

MirandaWood replied:
No problem - I also realise this PR diff is pretty hard to read because I rearranged existing code, so it looks to have a higher footprint than it really does. I'll just push the `EXPECTED_C_TAG` changes for this comment thread, thanks both :)

jeanmon commented on barretenberg/cpp/pil/vm2/alu.pil:
```diff
@@ -311,6 +286,47 @@ sel_is_u128 * sel_op_div * (1 - sel_err) * a_hi * b_hi = 0;
+// DIV & FDIV - non u128
+
+// FDIV note: I'm using (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff) to avoid having to set and check sel_is_u128 for FDIV, since we know it's FF => non u128
+pol DIV_OPS_NON_U128 = DIV_OPS * (IS_NOT_U128 + sel_is_ff - IS_NOT_U128 * sel_is_ff);
```
Why not defining:
`pol DIV_OPS_NON_U128 = sel_op_fdiv + sel_op_div * IS_NOT_U128`

MirandaWood replied:
This should be explained on the comment above, but I can reword it?
Essentially I don't use `sel_is_u128` for fdiv, we only need to know whether it is FF or not. This means for fdiv `IS_NOT_U128` (= 1 - sel_is_u128) may be incorrectly set to 1. If it is, I counter this by taking `IS_NOT_U128 * sel_is_ff`.
Instead, I could define `IS_NOT_U128 = 1 - sel_is_u128 + sel_is_ff` or something like that?
My idea was to avoid setting a bunch of columns and doing unnecessary `inv` checks, so I ignored `sel_is_u128`, but maybe it's better to be thorough?

jeanmon replied:
My suggested expression should achieve exactly the same, no?
I removed the influence of a badly set `sel_is_u128` for `sel_op_fdiv` and kept it only for `sel_op_div`.

MirandaWood replied:
I originally thought no because `sel_op_fdiv` does not guarantee `sel_is_ff` - it just means that if we have a non-FF input then we have a tag error. However `DIV_OPS_NON_U128` is only used when we _don't_ have an error, so it would work. Perhaps:
`pol DIV_OPS_NON_U128 =  (1 - sel_err) * (sel_op_fdiv + sel_op_div * IS_NOT_U128)`,
Then the main div relation could be just:
```
#[ALU_FDIV_DIV_NON_U128]
DIV_OPS_NON_U128 * (ib * ic - (ia - sel_op_div * helper1)) = 0;
```
What do you think?

jeanmon replied:
I like it, you can apply this.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/alu.test.cpp:
```diff
@@ -1016,6 +1021,208 @@ TEST_F(AluDivConstrainingTest, NegativeAluDivByZeroFFTagMismatch)
+
+// FDIV TESTS
+
+// Note: The test framework below converts all inputs to FF values to alllow for many happy path tests without adding
```
```suggestion
// Note: The test framework below converts all inputs to FF values to allow for many happy path tests without adding
```

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -97,6 +97,29 @@ MemoryValue Alu::div(const MemoryValue& a, const MemoryValue& b)
+    try {
+        MemoryValue c = a / b; // This will throw if the tags do not match or if we divide by 0.
+
+        if (a.get_tag() != MemoryTag::FF) {
```
Maybe we should replace this with an assertion. In the current way it is integrated this will never happen because the caller in execution will throw before reaching this code.

MirandaWood replied:
Yep - we can't reach this calling `execution.fdiv()` now, I thought the idea was just to continue to check on 'both sides' (as discussed in standup). I left this as is so we can test (in the simulation test) that a tag error is correctly permeated to the event, even though it's not the 'real' path. I'm happy to change this though!

jeanmon replied:
If it facilitates unit tests, I am fine, even though using a non/real code path is not so great. In any case, write a comment that a real path should never reaches it.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -97,6 +97,29 @@ MemoryValue Alu::div(const MemoryValue& a, const MemoryValue& b)
+
+        events.emit({ .operation = AluOperation::FDIV, .a = a, .b = b, .c = c });
+        return c;
+    } catch (const TagMismatchException& e) {
```
Same for this error, it should never happen.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/simulation/execution.test.cpp:
```diff
@@ -171,6 +171,20 @@ TEST_F(ExecutionSimulationTest, Div)
+    EXPECT_CALL(memory, set(3, MemoryValue::from<FF>(FF::modulus - 2)));
+    EXPECT_CALL(gas_tracker, consume_gas(Gas{ 0, 0 }));
+
+    execution.fdiv(context, 1, 2, 3);
```
Can you add a comment that the argument values 1 and 2 are completely arbitrary and are not checked in the various EXPECT_CALL()

MirandaWood replied:
I can - however this is also true for most of the other tests in this file, so maybe a comment at the top of the file would be better?

jeanmon replied:
Yes good idea.

jeanmon commented on barretenberg/cpp/src/barretenberg/vm2/tracegen/alu_trace.test.cpp:
```diff
@@ -631,6 +632,140 @@ TEST_F(AluTraceGenerationTest, TraceGenerationDivTagError)
 
+// FDIV TESTS
+
+// Note: The test framework below converts all inputs to FF values to alllow for many happy path tests without adding
```
```suggestion
// Note: The test framework below converts all inputs to FF values to allow for many happy path tests without adding
```

jeanmon reviewed (approved):
Great work overall. Please look at the comments before merging.

MirandaWood committed:
feat(avm)!: WIP ALU FDIV

## a / b = c, v.2: electric boogaloo

Adds field division opcode `FDIV` (somewhat following [vm1](https://github.com/AztecProtocol/aztec-packages/blame/bf7a99d8d8e3faa00c9395a17acf7f92b171a7a1/barretenberg/cpp/pil/avm/main.pil#L328)). Given fields `a` and `b`, shows that `a / b = c` by constraining `a = c * b`.

I re-use the existing `DIV` relation by forcing the remainder to be zero, and add the case of a non-FF input as a tag error. Also uses the existing divide by zero error plus tests!
